[김예랑] feature/carts

Yerang-Kim opened the pull request:
Add: carts app
add carts/views.py => get, post, delete API
add carts/tests.py => unit test => success case / error case

>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 카트 앱 작성
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- views.py GET => 특정 사용자의 카트 보여주기
- views.py POST => 특정 사용자가 특정 작품 카트애 담기
- views.py DELETE => 특정 사용자의 카트에서 특정 작품 빼기
- tests.py => 유닛테스트 (성공, 에러 케이스)

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)
-

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.
- [x] Git rebase와 squash를 했고 커밋 수가 하나 인것을 확인 했습니다.
- [x] 유닛 테스트를 구현 하였고 모든 유닛 테스트가 통과 된것을 확인 하였습니다.

Yerang-Kim commented:
<img width="1368" alt="스크린샷 2021-07-25 오후 6 02 21" src="https://user-images.githubusercontent.com/74804995/126894088-1f0d9073-d790-4aef-8ff1-9d16b681618c.png">
<img width="1352" alt="스크린샷 2021-07-25 오후 6 02 57" src="https://user-images.githubusercontent.com/74804995/126894090-ba31a757-4123-458e-95e1-206c2ac3264c.png">
<img width="1358" alt="스크린샷 2021-07-25 오후 6 04 29" src="https://user-images.githubusercontent.com/74804995/126894091-cbe845f2-111c-43f7-abae-634b6480d028.png">

Yerang-Kim commented:
통신 결과 캡쳐 같이 올립니닷

Kang-kyunghun commented on core/utils.py:
```diff
@@ -0,0 +1,53 @@
+
+    return wrapper
+
+def query_debugger(func):
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        reset_queries()
+        number_of_start_queries = len(connection.queries)
+
+        start  = time.perf_counter()
+        result = func(*args, **kwargs)
+        end    = time.perf_counter()
+
+        number_of_end_queries   = len(connection.queries)
+        print(f"-------------------------------------------------------------------")
+        print(f"Function : {func.__name__}")
+        print(f"Number of Queries : {number_of_end_queries-number_of_start_queries}")
+        print(f"Finished in : {(end - start):.2f}s")
+        print(f"-------------------------------------------------------------------")
+        return result
+    return wrapper
```
디버거가 포함된 코드는 실제 서비스의 속도를 느려지게 하기 때문에 디버거 같은 경우에는 테스트에서만 사용되어야 합니다. 공통으로 사용될 코드가 모이는 github에서는 제외 되어야 합니다.

Yerang-Kim replied:
넵 깃헙에 올릴 때에는 제외하고 올리겠습니다

Kang-kyunghun commented on carts/views.py:
```diff
@@ -1,3 +1,90 @@
+        
+        except User.DoesNotExist:
+            return JsonResponse({'message': 'USER_DOES_NOT_EXISTS'}, status=400)
+        except Exception as error:
+            return JsonResponse({'message': error}, status=400)
```
다음과 같은 예외 처리는 몇 가지 문제점이 있습니다.
1. 위에서 처리한 예외처리들이 모두 해당 예외처리에 걸릴 수 있습니다. 
2. as error을 통해 에러 메시지를 만들어서 어떤 에러인지 특정 지을 수는 있지만, 이와 같이 처리 할 경우 프론트에서 어디 파일 몇 번째 줄에서 에러가 났는지 확일 할 수 있습니다. 이럴 경우 보안에 문제가 되기 때문에 피해야 하는 방법 입니다.

Yerang-Kim replied:
넵 반영하겠습니다.

Kang-kyunghun commented on carts/views.py:
```diff
@@ -1,3 +1,90 @@
+        try:
+            user  = request.user
+            items = Cart.objects.filter(user=user)
+            arts  = Art.objects.select_related('artist', 'size').filter(id__in=[item.art_id for item in items])
+            
```
cart를 통해서 충분히 art들을 가져올 수 있을 것 같은데, 한 번 고민해 보세요.

Yerang-Kim replied:
나름 고민하여 반영해보았습니다.

Kang-kyunghun commented on carts/views.py:
```diff
@@ -1,3 +1,90 @@
+                for art in arts
+            ]
+
+            total_price = int(arts.aggregate(sum=Sum('price'))['sum'])
```
aggregate는 arts 데이터 가져 올 때 한 번에 가져와서 사용하면 좋을 것 같습니다.

Kang-kyunghun commented on carts/views.py:
```diff
@@ -1,3 +1,90 @@
+            ]
+
+            total_price = int(arts.aggregate(sum=Sum('price'))['sum'])
+            count       = len(arts)
```
django ORM 중에 .count() method 한 번 활용해 보세요.

Yerang-Kim replied:
넵 !!!

Kang-kyunghun commented on carts/views.py:
```diff
@@ -1,3 +1,90 @@
+
+    @query_debugger
+    @login_required
+    def delete(self, request):
+        try:
+            art_id = request.GET.get('art-id')
```
path parameter을 활용해 보세요~

Yerang-Kim replied:
넵 반영햇습니다

Kang-kyunghun commented on carts/views.py:
```diff
@@ -1,3 +1,90 @@
+            art    = Art.objects.get(id=art_id)
+            user   = request.user
+
+            if not Cart.objects.filter(art=art, user=user).exists():
+                return JsonResponse({'message': 'INVALID_REQUEST'}, status=400)
```
404 NOT FOUND가 조금 더 적절해 보입니다.

Yerang-Kim replied:
넵 반영햇습니다

Kang-kyunghun commented on carts/views.py:
```diff
@@ -1,3 +1,90 @@
+            art    = Art.objects.get(id=art_id)
+            user   = request.user
+
+            item, is_created = Cart.objects.get_or_create(art=art, user=user)
```
get_or_create를 사용한 이유를 코멘트로 남겨 주세요.

Yerang-Kim replied:
저도 왜 썼는지 모르겠습니당... 그래서 코드 바꿨습니닷.....

Kang-kyunghun commented:
aggregate 관련 리뷰는 지웠습니다. 제가 annotate랑 착각 했습니다.

Yerang-Kim commented:
리뷰 반영하여 푸쉬했습니다

Kang-kyunghun commented on carts/views.py:
```diff
@@ -1,3 +1,82 @@
+    @login_required
+    def delete(self, request, art_id):
+        try:
+            art    = Art.objects.get(id=art_id)
```
```suggestion
            
```
art는 제거하고 art_id로 처리하면 될 것 같아요.

Yerang-Kim replied:
넵 수정햇습니다

Kang-kyunghun commented on carts/views.py:
```diff
@@ -1,3 +1,82 @@
+            art    = Art.objects.get(id=art_id)
+            user   = request.user
+
+            if not Cart.objects.filter(art=art, user=user).exists():
```
```suggestion
            if not Cart.objects.filter(art_id=art_id, user=user).exists():
```

Kang-kyunghun commented on carts/views.py:
```diff
@@ -1,3 +1,82 @@
+            if not Cart.objects.filter(art=art, user=user).exists():
+                return JsonResponse({'message': 'NOT_FOUND'}, status=404)
+
+            Cart.objects.get(art=art, user=user).delete()
```
```suggestion
            Cart.objects.get(art_id=art_id, user=user).delete()
```

Kang-kyunghun commented on core/utils.py:
```diff
@@ -0,0 +1,34 @@
+        except KeyError:
+            return JsonResponse({'message': 'KEY_ERROR'}, status=400)
+        except Exception as error:
+            return JsonResponse({'message': error}, status=400)
```
제거해 주세요.

Yerang-Kim replied:
넵 제거햇습니다!

grondin0425 committed:
Add: carts app
add carts/views.py => get, post, delete API
add carts/tests.py => unit test => success case / error case

Yerang-Kim commented:
경훈님!!
커밋이 두개가 되엇습니다..
그리고 컨플릭트가 사라졋습니다
로컬에서 제가 리베이스 컨티뉴 이런거 햇어서 그런거 같습니닷... 죄송합니다 ㅜㅜ
이따가 시간 되실때 한번 봐주시면 감사하겟습니다!

Kang-kyunghun commented:
LGTM~!!
